File: sdk/python/moltspeak/session.py
"""
MoltSpeak Session management
"""
import time
import secrets
from dataclasses import dataclass, field
from typing import Optional, Dict, List, Any


@dataclass
class Session:
    """
    Represents an authenticated session between two agents.
    
    Sessions are established after successful handshake and provide:
    - Session ID for message correlation
    - Shared session key (from key exchange)
    - Negotiated capabilities
    - Expiration tracking
    """
    session_id: str
    local_agent: str
    remote_agent: str
    remote_org: str
    remote_public_key: str
    session_key: Optional[bytes] = None  # Derived from key exchange
    capabilities: List[str] = field(default_factory=list)
    extensions: List[str] = field(default_factory=list)
    created_at: int = field(default_factory=lambda: int(time.time() * 1000))
    expires_at: Optional[int] = None  # Unix ms
    last_activity: int = field(default_factory=lambda: int(time.time() * 1000))
    message_count: int = 0
    
    @classmethod
    def create(
        cls,
        local_agent: str,
        remote_agent: str,
        remote_org: str,
        remote_public_key: str,
        ttl_seconds: int = 3600
    ) -> "Session":
        """Create a new session"""
        now = int(time.time() * 1000)
        return cls(
            session_id=secrets.token_hex(16),
            local_agent=local_agent,
            remote_agent=remote_agent,
            remote_org=remote_org,
            remote_public_key=remote_public_key,
            expires_at=now + (ttl_seconds * 1000),
        )
    
    def is_expired(self) -> bool:
        """Check if session has expired"""
        if self.expires_at is None:
            return False
        return int(time.time() * 1000) > self.expires_at
    
    def is_valid(self) -> bool:
        """Check if session is valid for use"""
        return not self.is_expired()
    
    def touch(self) -> None:
        """Update last activity timestamp"""
        self.last_activity = int(time.time() * 1000)
        self.message_count += 1
    
    def has_capability(self, capability: str) -> bool:
        """Check if remote agent has a capability"""
        return capability in self.capabilities
    
    def extend(self, ttl_seconds: int = 3600) -> None:
        """Extend session expiration"""
        self.expires_at = int(time.time() * 1000) + (ttl_seconds * 1000)


class SessionManager:
    """
    Manage multiple sessions.
    
    Tracks active sessions, handles expiration, limits per-peer sessions.
    """
    
    def __init__(self, max_sessions: int = 100, default_ttl: int = 3600):
        self._sessions: Dict[str, Session] = {}
        self._by_remote: Dict[str, List[str]] = {}  # remote_agent -> [session_ids]
        self.max_sessions = max_sessions
        self.default_ttl = default_ttl
    
    def create(
        self,
        local_agent: str,
        remote_agent: str,
        remote_org: str,
        remote_public_key: str,
        ttl_seconds: Optional[int] = None
    ) -> Session:
        """Create and store a new session"""
        # Clean up expired sessions first
        self._cleanup_expired()
        
        # Check limits
        if len(self._sessions) >= self.max_sessions:
            # Remove oldest session
            oldest = min(self._sessions.values(), key=lambda s: s.last_activity)
            self.remove(oldest.session_id)
        
        session = Session.create(
            local_agent=local_agent,
            remote_agent=remote_agent,
            remote_org=remote_org,
            remote_public_key=remote_public_key,
            ttl_seconds=ttl_seconds or self.default_ttl,
        )
        
        self._sessions[session.session_id] = session
        
        if remote_agent not in self._by_remote:
            self._by_remote[remote_agent] = []
        self._by_remote[remote_agent].append(session.session_id)
        
        return session
    
    def get(self, session_id: str) -> Optional[Session]:
        """Get session by ID"""
        session = self._sessions.get(session_id)
        if session and session.is_expired():
            self.remove(session_id)
            return None
        return session
    
    def get_for_remote(self, remote_agent: str) -> Optional[Session]:
        """Get an active session for a remote agent"""
        session_ids = self._by_remote.get(remote_agent, [])
        for sid in session_ids:
            session = self.get(sid)
            if session and session.is_valid():
                return session
        return None
    
    def remove(self, session_id: str) -> None:
        """Remove a session"""
        session = self._sessions.pop(session_id, None)
        if session and session.remote_agent in self._by_remote:
            self._by_remote[session.remote_agent] = [
                sid for sid in self._by_remote[session.remote_agent]
                if sid != session_id
            ]
    
    def _cleanup_expired(self) -> None:
        """Remove all expired sessions"""
        expired = [
            sid for sid, session in self._sessions.items()
            if session.is_expired()
        ]
        for sid in expired:
            self.remove(sid)
    
    def active_count(self) -> int:
        """Count active sessions"""
        self._cleanup_expired()
        return len(self._sessions)
    
    def list_sessions(self) -> List[Session]:
        """List all active sessions"""
        self._cleanup_expired()
        return list(self._sessions.values())

File: sdk/js/src/session.ts
/**
 * MoltSpeak Session management
 */

import type { SessionState } from './types';

/**
 * Session between two agents
 */
export class Session implements SessionState {
  sessionId: string;
  localAgent: string;
  remoteAgent: string;
  remoteOrg: string;
  remotePublicKey: string;
  sessionKey?: Uint8Array;
  capabilities: string[];
  extensions: string[];
  createdAt: number;
  expiresAt: number | null;
  lastActivity: number;
  messageCount: number;

  constructor(params: {
    sessionId: string;
    localAgent: string;
    remoteAgent: string;
    remoteOrg: string;
    remotePublicKey: string;
    capabilities?: string[];
    extensions?: string[];
    ttlSeconds?: number;
  }) {
    const now = Date.now();
    this.sessionId = params.sessionId;
    this.localAgent = params.localAgent;
    this.remoteAgent = params.remoteAgent;
    this.remoteOrg = params.remoteOrg;
    this.remotePublicKey = params.remotePublicKey;
    this.capabilities = params.capabilities || [];
    this.extensions = params.extensions || [];
    this.createdAt = now;
    this.expiresAt = params.ttlSeconds ? now + params.ttlSeconds * 1000 : null;
    this.lastActivity = now;
    this.messageCount = 0;
  }

  /**
   * Create a new session
   */
  static create(
    localAgent: string,
    remoteAgent: string,
    remoteOrg: string,
    remotePublicKey: string,
    ttlSeconds = 3600
  ): Session {
    return new Session({
      sessionId: crypto.randomUUID(),
      localAgent,
      remoteAgent,
      remoteOrg,
      remotePublicKey,
      ttlSeconds,
    });
  }

  /**
   * Check if session has expired
   */
  isExpired(): boolean {
    if (this.expiresAt === null) return false;
    return Date.now() > this.expiresAt;
  }

  /**
   * Check if session is valid
   */
  isValid(): boolean {
    return !this.isExpired();
  }

  /**
   * Update activity timestamp
   */
  touch(): void {
    this.lastActivity = Date.now();
    this.messageCount++;
  }

  /**
   * Check if remote has capability
   */
  hasCapability(capability: string): boolean {
    return this.capabilities.includes(capability);
  }

  /**
   * Extend session expiration
   */
  extend(ttlSeconds = 3600): void {
    this.expiresAt = Date.now() + ttlSeconds * 1000;
  }
}

/**
 * Manage multiple sessions
 */
export class SessionManager {
  private sessions: Map<string, Session> = new Map();
  private byRemote: Map<string, string[]> = new Map();
  maxSessions: number;
  defaultTTL: number;

  constructor(maxSessions = 100, defaultTTL = 3600) {
    this.maxSessions = maxSessions;
    this.defaultTTL = defaultTTL;
  }

  /**
   * Create and store a new session
   */
  create(
    localAgent: string,
    remoteAgent: string,
    remoteOrg: string,
    remotePublicKey: string,
    ttlSeconds?: number
  ): Session {
    // Cleanup expired first
    this.cleanupExpired();

    // Check limits
    if (this.sessions.size >= this.maxSessions) {
      // Remove oldest
      let oldest: Session | null = null;
      for (const session of this.sessions.values()) {
        if (!oldest || session.lastActivity < oldest.lastActivity) {
          oldest = session;
        }
      }
      if (oldest) {
        this.remove(oldest.sessionId);
      }
    }

    const session = Session.create(
      localAgent,
      remoteAgent,
      remoteOrg,
      remotePublicKey,
      ttlSeconds ?? this.defaultTTL
    );

    this.sessions.set(session.sessionId, session);

    // Track by remote agent
    const remoteList = this.byRemote.get(remoteAgent) || [];
    remoteList.push(session.sessionId);
    this.byRemote.set(remoteAgent, remoteList);

    return session;
  }

  /**
   * Get session by ID
   */
  get(sessionId: string): Session | null {
    const session = this.sessions.get(sessionId);
    if (session && session.isExpired()) {
      this.remove(sessionId);
      return null;
    }
    return session || null;
  }

  /**
   * Get active session for remote agent
   */
  getForRemote(remoteAgent: string): Session | null {
    const sessionIds = this.byRemote.get(remoteAgent) || [];
    for (const sid of sessionIds) {
      const session = this.get(sid);
      if (session && session.isValid()) {
        return session;
      }
    }
    return null;
  }

  /**
   * Remove a session
   */
  remove(sessionId: string): void {
    const session = this.sessions.get(sessionId);
    if (session) {
      this.sessions.delete(sessionId);
      const remoteList = this.byRemote.get(session.remoteAgent);
      if (remoteList) {
        this.byRemote.set(
          session.remoteAgent,
          remoteList.filter((id) => id !== sessionId)
        );
      }
    }
  }

  /**
   * Cleanup expired sessions
   */
  private cleanupExpired(): void {
    for (const [id, session] of this.sessions) {
      if (session.isExpired()) {
        this.remove(id);
      }
    }
  }

  /**
   * Count active sessions
   */
  activeCount(): number {
    this.cleanupExpired();
    return this.sessions.size;
  }

  /**
   * List all active sessions
   */
  list(): Session[] {
    this.cleanupExpired();
    return Array.from(this.sessions.values());
  }
}
